modules/renter/contractor/formcontract.go

```diff
@@ -3,311 +3,59 @@ package contractor
-	}
-	revisionTxn.TransactionSignatures[0].Signature = encodedSig[:]
-
-	// Send acceptance and signatures
-	if err := modules.WriteNegotiationAcceptance(conn); err != nil {
-		return Contract{}, errors.New("couldn't send transaction acceptance: " + err.Error())
-	}
-	if err := encoding.WriteObject(conn, addedSignatures); err != nil {
-		return Contract{}, errors.New("couldn't send added signatures: " + err.Error())
-	}
-	if err := encoding.WriteObject(conn, revisionTxn.TransactionSignatures[0]); err != nil {
-		return Contract{}, errors.New("couldn't send revision signature: " + err.Error())
-	}
-
-	// Read the host acceptance and signatures.
-	err = modules.ReadNegotiationAcceptance(conn)
-	if err != nil {
-		return Contract{}, errors.New("host did not accept our signatures: " + err.Error())
-	}
-	var hostSigs []types.TransactionSignature
-	if err := encoding.ReadObject(conn, &hostSigs, 2e3); err != nil {
-		return Contract{}, errors.New("couldn't read the host's signatures: " + err.Error())
-	}
-	for _, sig := range hostSigs {
-		txnBuilder.AddTransactionSignature(sig)
-	}
-	var hostRevisionSig types.TransactionSignature
-	if err := encoding.ReadObject(conn, &hostRevisionSig, 2e3); err != nil {
-		return Contract{}, errors.New("couldn't read the host's revision signature: " + err.Error())
-	}
-	revisionTxn.TransactionSignatures = append(revisionTxn.TransactionSignatures, hostRevisionSig)
-
-	// Construct the final transaction.
-	txn, parentTxns = txnBuilder.View()
-	txnSet = append(parentTxns, txn)
-
-	// submit to blockchain
-	err = tpool.AcceptTransactionSet(txnSet)
-	if err == modules.ErrDuplicateTransactionSet {
-		// as long as it made it into the transaction pool, we're good
-		err = nil
-	}
-	if err != nil {
-		return Contract{}, err
-	}
-
-	// calculate contract ID
-	fcid := txnSet[len(txnSet)-1].FileContractID(0)
-
-	// create host contract
-	contract := Contract{
-		IP:              host.NetAddress,
-		ID:              fcid,
-		FileContract:    fc,
-		LastRevision:    initRevision,
-		LastRevisionTxn: revisionTxn,
-		SecretKey:       ourSK,
-	}
-
-	return contract, nil
-}
-
-// newContract negotiates an initial file contract with the specified host
-// and returns a Contract. The contract is also saved by the HostDB.
-func (c *Contractor) newContract(host modules.HostDBEntry, filesize uint64, endHeight types.BlockHeight) (Contract, error) {
+// newContract negotiates an initial file contract with the specified host,
+// saves it, and returns it.
+func (c *Contractor) newContract(host modules.HostDBEntry, filesize uint64, duration types.BlockHeight) (proto.Contract, error) {
 	// reject hosts that are too expensive
 	if host.StoragePrice.Cmp(maxStoragePrice) > 0 {
-		return Contract{}, errTooExpensive
+		return proto.Contract{}, errTooExpensive
 	}
 
-	// get an address to use for negotiation
+	// TODO: move this outside this function?
 	c.mu.Lock()
+	// get an address to use for negotiation
```

Comment:
come to think of it, `cachedAddress` may no longer be necessary, not that we aren't creating a new contract for each file. Worst case, you'll get a handful of wasted addresses when you try to set an allowance.